projecteurapp.cc

```diff
@@ -139,6 +139,11 @@ ProjecteurApplication::ProjecteurApplication(int &argc, char **argv)
       window->setFlags(window->flags() & ~Qt::SplashScreen);
       window->setFlags(window->flags() | Qt::ToolTip);
 
+      // A hack to automatically update border changes
```

Comment:
Why this hack?

Reply:
I was searching for an event of that is triggered every time  border properties changes. I was not able to find a suitable event so I used `onVisibleChange` for that. 

Hence, this hack try to trigger the event so that the borders are updated as per the settings.

Reply:
You can solve this more elegantly without the hack in the projecteurapp.cc, in the main.qml for example for your borderLoader:

```
        Loader {
            id: borderShapeLoader
            anchors.centerIn: centerRect
            width: centerRect.width;  height: width
            sourceComponent: Qt.createComponent(Settings.spotShape)

            onSourceComponentChanged: {
                borderShapeLoader.item.visible = true;
                borderShapeLoader.item.color = "transparent";
                borderShapeLoader.item.opacity = Qt.binding(function() {return 1-Settings.shadeOpacity});
                borderShapeLoader.item.border.width = Qt.binding(function() {return Settings.borderSize/100*borderShapeLoader.width});
                borderShapeLoader.item.border.color = Qt.binding(function() {return Settings.borderColor});
            }
            visible: Settings.showBorder
        }
```

Reply:
This will create bindings in Javascript as soon as the sourceComponent changes are automatically updated by qml due to the binding.

Reply:
It does not work until user change the `spotShape`. If the user is just changing the border properties then border do not get updated.

The problem here is the component (`Settings.spotShape`) is not getting changed so QML do not trigger the `onSourceComponentChanged` automatically whenever the border properties changes.

Reply:
I just tried it locally with your branch - works for me. Because sourceComponentChanged is always executed... just put a `console.log("hello")` output in there and try it

Reply:
I am saying the same thing - the code works when user change the component (`spotShape`).

To reproduce the error -- first set `spotShape` and then check the spot using test button. After that change the border color and size (but not anything else). By pressing test button you can see the border fail to update to new color and size.

Reply:
Definitly works for me - it does not matter what I change and not change in what order.. If the binding is set up correctly no workaround is necessary. Does not depend on the device either - only testing with the test button also. If you don't see outputs for onSourceComponentChanged if you have a log output there then sth else must be wrong.

But the following was still missing `if (!borderShapeLoader.item.border) return;` (otherwise shapes that don't have border property will result in qml error messages:

The following one works for me with the test button always, no matter what you change in what order.

```qml
        Loader {
            id: borderShapeLoader
            anchors.centerIn: centerRect
            width: centerRect.width;  height: width
            sourceComponent: Qt.createComponent(Settings.spotShape)

            onSourceComponentChanged: {
                if (!borderShapeLoader.item.border) return;
                borderShapeLoader.item.visible = true;
                borderShapeLoader.item.color = "transparent";
                borderShapeLoader.item.opacity = Qt.binding(function() {return 1-Settings.shadeOpacity});
                borderShapeLoader.item.border.width = Qt.binding(function() {return Settings.borderSize/100*borderShapeLoader.width});
                borderShapeLoader.item.border.color = Qt.binding(function() {return Settings.borderColor});
            }
            visible: Settings.showBorder
        }
```

Reply:
Here is my complete main.qml file: https://gist.github.com/jahnf/198433b5b81b2367d8f8cdc4c4f2eb42

Reply:
I have found the error on my side. I missed the `Qt.binding(` binding part. I confirm it is working.

Reply:
I have pushed the changes. Thanks for the suggestions.